byterun/minor_gc.c

```diff
@@ -190,9 +246,29 @@ void caml_oldify_one (value v, value *p)
       *p = Field (v, 0);  /*  then forward pointer is first field. */
     }else{
       tag = Tag_hd (hd);
+#ifdef NATIVE_CODE_AND_NO_NAKED_POINTERS
+      if (tag == Closure_tag) {
+        uintnat first_part = caml_distance_to_closure_environment(v);
+        sz = Wosize_hd (hd);
+        result = caml_alloc_shr_preserving_profinfo (sz, tag, hd);
+        for (i = 0; i < sz; i++) {
+          value field = Field (v, i);
+          if (i >= first_part && Is_block (field) && Is_young (field)) {
+            /* This seems most unlikely to cause a stack overflow. */
```

Comment:
Hughes-style difference lists will cause this to stack overflow:

```
type 'a dlist = 'a list -> 'a list
let singleton x = fun tail -> x :: tail
let concat a b = fun tail -> a (b tail)
```
A long list consists of deeply nested closures.